### PR TITLE
INTMDB-416-database-user Adding x509 User

### PR DESCRIPTION
**Pull request by ParthasarathyVaradhan**
## Description

Please include a summary of the fix/feature/change, including any relevant motivation and context.

Link to any related issue(s): 

## Type of change:

- [✅] Bug fix (non-breaking change which fixes an issue)
- [✅ ] New feature (non-breaking change which adds functionality)
-    [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [✅ ] This change requires a documentation update

## Required Checklist:

- [✅] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [✅] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [✅] I have run `make fmt` and formatted my code

## Further comments

INTMDB-416-database-user

New Change:
Code for new X509 database-user type.

Bug Fixes And Enhancements: 
Scopes updates during update operation.
Validation of input parameters.

##Testing templates:

The Cloudformation template below
cfn-resources/database-user/test/databaseuser.sample-template.json
tests creation of 4 different types of user X509TypeUser, ScramTypeUser, LDAPTypeUser, AWSIAMTypeUser

### CODE REVIEW

**Comment by Zuhairahmed** [comment]
Add link to DatabaseUser Example

File: cfn-resources/database-user/README.md [hunk 1]
```diff
@@ -56,4 +56,5 @@ source <(${repo_root}/quickstart-mongodb-atlas/scripts/export-mongocli-config.py
 ${repo_root}/quickstart-mongodb-atlas/scripts/launch-x-quickstart.sh ${repo_root}/cfn-resources/database-user/test/databaseuser.sample-template.yaml SampleAccessList1 ParameterKey=ProjectId,ParameterValue=<YOUR_PROJECT_ID> ParameterKey=Username,ParameterValue=<USERNAME> ParameterKey=Password,ParameterValue=<PWD> 
 ```
 
-
+For more information please refer API Documentation,
+https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Database-Users
```
Comment (Zuhairahmed) [inline cfn-resources/database-user/README.md:60]: Copy same format as footer in MongoDB::Atlas::Project reference to Documentation

File: cfn-resources/database-user/README.md [hunk 2]
```diff
@@ -56,4 +56,5 @@ source <(${repo_root}/quickstart-mongodb-atlas/scripts/export-mongocli-config.py
 ${repo_root}/quickstart-mongodb-atlas/scripts/launch-x-quickstart.sh ${repo_root}/cfn-resources/database-user/test/databaseuser.sample-template.yaml SampleAccessList1 ParameterKey=ProjectId,ParameterValue=<YOUR_PROJECT_ID> ParameterKey=Username,ParameterValue=<USERNAME> ParameterKey=Password,ParameterValue=<PWD> 
 ```
```
Comment (Zuhairahmed) [inline cfn-resources/database-user/README.md:58]: Add new section + link to DatabaseUser Example

File: cfn-resources/database-user/docs/README.md [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # MongoDB::Atlas::DatabaseUser
 
-The databaseUsers resource lets you retrieve, create and modify the MongoDB users in your cluster. Each user has a set of roles that provide access to the project’s databases. A user’s roles apply to all the clusters in the project: if two clusters have a products database and a user has a role granting read access on the products database, the user has that access on both clusters.
+Returns, adds, edits, and removes database users.
```
Comment (Zuhairahmed) [inline cfn-resources/database-user/docs/README.md:4]: Why do we have 2 README files for same resource?

File: cfn-resources/database-user/test/databaseuser.sample-template.json [hunk 4]
```diff
@@ -0,0 +1,131 @@
+{
+  "AWSTemplateFormatVersion": "2010-09-09",
+  "Description": "Returns all database users that belong to the specified project. To use this resource, the requesting API Key must have the Project Read Only role. This resource doesn't require the API Key to have an Access List.",
```
Comment (Zuhairahmed) [inline cfn-resources/database-user/test/databaseuser.sample-template.json:3]: Is this file new Example? If so should we delete https://github.com/mongodb/mongodbatlas-cloudformation-resources/blob/master/examples/database-user/user.json ?

**Review by Zuhairahmed — COMMENTED** [review]
Mainly doc comments below

File: cfn-resources/database-user/test/databaseuser.sample-template.json (hunk 4 above)
Comment (ParthasarathyVaradhan) [reply cfn-resources/database-user/test/databaseuser.sample-template.json:3]: As far as the https://github.com/mongodb/mongodbatlas-cloudformation-resources/blob/master/examples/README.md is suggesting, 

"These examples are deprecated and will soon be removed from the code base in favor of the test/project.sample-cfn-request.json and test/project.sample-template.yaml examples contain in the source folder for each Resource Type."
Following this comment I have kept the file in ./test/databaseuser.sample-template.json.
I will delete this https://github.com/mongodb/mongodbatlas-cloudformation-resources/blob/master/examples/database-user/user.json file

Comment (ParthasarathyVaradhan) [reply cfn-resources/database-user/test/databaseuser.sample-template.json:3]: Handled in [6d075bc20598253c789ce5f833954a06a5fe750a](https://github.com/mongodb/mongodbatlas-cloudformation-resources/pull/132/commits/6d075bc20598253c789ce5f833954a06a5fe750a)

File: cfn-resources/database-user/README.md (hunk 2 above)
Comment (ParthasarathyVaradhan) [reply cfn-resources/database-user/README.md:62]: Handled in [6d075bc20598253c789ce5f833954a06a5fe750a](https://github.com/mongodb/mongodbatlas-cloudformation-resources/pull/132/commits/6d075bc20598253c789ce5f833954a06a5fe750a)

File: cfn-resources/database-user/README.md (hunk 1 above)
Comment (ParthasarathyVaradhan) [reply cfn-resources/database-user/README.md]: Handled in [6d075bc20598253c789ce5f833954a06a5fe750a](https://github.com/mongodb/mongodbatlas-cloudformation-resources/pull/132/commits/6d075bc20598253c789ce5f833954a06a5fe750a)

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 5]
```diff
@@ -15,11 +17,27 @@ func setup() {
...
+// function to validate inputs to all actions
+func validateModel(fields []string, model *Model) *handler.ProgressEvent {
+	return validator.ValidateModel(fields, model)
+}
+
+var CreateRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "DatabaseName", "ProjectId", "Roles", "Username"}
+var ReadRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "ProjectId", "DatabaseName", "Username"}
+var UpdateRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "DatabaseName", "ProjectId", "Roles", "Username"}
+var DeleteRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "ProjectId", "DatabaseName", "Username"}
+var ListRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "ProjectId"}
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:29]: Please define


```
const (
	publicKey    = "ApiKeys.PublicKey"
	privateKey   = "ApiKeys.PrivateKey"
	databaseName = "DatabaseName"
	projectId    = "ProjectId"
	roles        = "Roles"
	username     = "Username"
)
```
then 
```
var CreateRequiredFields = []string{publicKey, privateKey, databaseName, projectId, roles, username}
```

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 6]
```diff
@@ -15,11 +17,27 @@ func setup() {
...
+
 // Create handles the Create event from the Cloudformation service.
 func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
 	setup()
 	log.Debugf(" currentModel: %#+v, prevModel: %#+v", currentModel, prevModel)
 
+	modelValidation := validateModel(CreateRequiredFields, currentModel)
+	if modelValidation != nil {
+		return *modelValidation, nil
+	}
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:39]:
```suggestion
	if modelValidation := validateModel(CreateRequiredFields, currentModel); modelValidation != nil {
		return *modelValidation, nil
	}
```

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 7]
```diff
@@ -94,41 +116,22 @@ func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler
 		currentModel.Password = &s
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:116]:
```suggestion
		currentModel.Password = pointy.String("")
```

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 8]
```diff
@@ -94,41 +116,22 @@ func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler
...
-		Username:     *currentModel.Username,
-		Password:     *currentModel.Password,
-		DatabaseName: *currentModel.DatabaseName,
-		Labels:       labels,
-		Scopes:       scopes,
-		LDAPAuthType: *currentModel.LdapAuthType,
-		AWSIAMType:   *currentModel.AWSIAMType,
+	if (currentModel.X509Type != &none) || (currentModel.DeleteAfterDate == nil) {
+		s := ""
+		currentModel.DeleteAfterDate = &s
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:121]:
```suggestion
		currentModel.DeleteAfterDate = pointy.String("")
```

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 9]
```diff
@@ -94,41 +116,22 @@ func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler
...
 	}
 
-	/*
-		projectResID := &util.ResourceIdentifier{
-			ResourceType: "Project",
-			ResourceID:   groupID,
-		}
-		resourceID := util.NewResourceIdentifier("DBUser", user.Username, projectResID)
+	log.Printf("Check Delete after date here::???????")
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:124]: Please, remove this comment

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 10]
```diff
@@ -82,9 +100,13 @@ func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler
...
 
+	if currentModel.X509Type == nil {
+		currentModel.X509Type = &none
+	}
+
 	if currentModel.Password == nil {
-		if (currentModel.LdapAuthType == &none) && (currentModel.AWSIAMType == &none) {
-			err := fmt.Errorf("Password cannot be empty if not LDAP or IAM: %v", currentModel)
+		if (currentModel.LdapAuthType == &none) && (currentModel.AWSIAMType == &none) && (currentModel.X509Type == &none) {
+			err := fmt.Errorf("Password cannot be empty if not LDAP or IAM or X509: %v", currentModel)
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:109]:
```suggestion
			err := fmt.Errorf("Password cannot be empty if LDAP, IAM or X509 are not provided: %v", currentModel)
```

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 11]
```diff
@@ -142,6 +145,12 @@ func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler
 // Read handles the Read event from the Cloudformation service.
 func Read(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
 	setup()
+
+	modelValidation := validateModel(ReadRequiredFields, currentModel)
+	if modelValidation != nil {
+		return *modelValidation, nil
+	}
+
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:153]: same as above

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 12]
```diff
@@ -258,41 +265,63 @@ func Update(req handler.Request, prevModel *Model, currentModel *Model) (handler
 		}
 		labels = append(labels, label)
 	}
+	log.Debugf("Update labels: %#+v", labels)
 
-	if currentModel.LdapAuthType == nil {
-		currentModel.LdapAuthType = new(string)
+	var scopes []mongodbatlas.Scope
+	for i, _ := range currentModel.Scopes {
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:271]:
```suggestion
	for i:= range currentModel.Scopes {
```

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 13]
```diff
@@ -258,41 +265,63 @@ func Update(req handler.Request, prevModel *Model, currentModel *Model) (handler
...
+	log.Printf("Check Delete after date here::???????")
+	spew.Dump(currentModel)
+
+	dbu := getDbUser(roles, groupID, currentModel, labels, scopes)
+	log.Debugf("dbu:%+v", dbu)
+	_, resp, err := client.DatabaseUsers.Update(context.Background(), groupID, *currentModel.Username, dbu)
+
+	log.Debugf("Update resp:%+v", resp)
+	if err != nil {
+		return progress_events.GetFailedEventByResponse(fmt.Sprintf("Error getting resource : %s", err.Error()),
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:323]: This logic is the same for the create. Can we extract it in a function and reuse it?

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 14]
```diff
@@ -302,10 +331,31 @@ func Update(req handler.Request, prevModel *Model, currentModel *Model) (handler
...
+
 // Delete handles the Delete event from the Cloudformation service.
 func Delete(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
 	setup()
 	log.Debugf("Create req:%+v, prevModel:%s, currentModel:%s", req, spew.Sdump(prevModel), spew.Sdump(currentModel))
+
+	modelValidation := validateModel(DeleteRequiredFields, currentModel)
+	if modelValidation != nil {
+		return *modelValidation, nil
+	}
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:358]: same as above

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 15]
```diff
@@ -347,6 +385,11 @@ func Delete(req handler.Request, prevModel *Model, currentModel *Model) (handler
 func List(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
 	setup()
 
+	modelValidation := validateModel(ListRequiredFields, currentModel)
+	if modelValidation != nil {
+		return *modelValidation, nil
+	}
+
```
Comment (andreaangiolillo) [inline cfn-resources/database-user/cmd/resource/resource.go:392]: same as above

**Review by Zuhairahmed — APPROVED** [review]
LGTM!

**Review by govardhanpagidi — APPROVED** [review]
LGTM

**Review by andreaangiolillo — APPROVED** [review]

**Pull request merged**